### DISCUSSION TITLE
Config for assembled pipeline ignores sourced components config structure.

### DISCUSSION THREAD

**Discussion by narayanacharya6** [explosion/spaCy#8278] [Help: Coding & Implementations]
## How to reproduce the behaviour
1. Train one (or multiple) custom components, say NER, and save it to a local directory of choice. The `ner.trainer.cfg` is a pretty standard cfg one sourced from the spaCy website.
`python -m spacy train "ner.trainer.cfg" --paths.train "./data/train-1.txt" --paths.dev "./data/test-1.txt"  -c "custom.py" --output "./model_store/ner_1/"`

`python -m spacy train "ner.trainer.cfg" --paths.train "./data/train-2.txt" --paths.dev "./data/test-2.txt"  -c "custom.py" --output "./model_store/ner_2/"`

2. Assemble a different pipeline that sources the above-trained component(s). 
`python -m spacy assemble "assemble.cfg" "./model_store/assembled/" --source.ner_1 ./model_store/ner_1/ --source.ner_2 ./model_store/ner_2/ -c custom_assemble.py`

The `assemble.cfg` looks something like this:
```
[source]
ner_1 = null
ner_2 = null

[system]
gpu_allocator = null

[nlp]
lang = "en"
pipeline = ["ner_1", "ner_2"]
batch_size = 256
disabled = []
before_creation = null
after_creation = null
after_pipeline_creation = null
tokenizer = {"@tokenizers":"spacy.Tokenizer.v1"}

[components]

[components.ner_1]
source = ${source.ner_1}
component = "ner"

[components.ner]
source = ${source.ner_model}
component = "ner"
```

3. Check the assembled model config under `./model_store/assembled/config.cfg`. The sourced NER models config is such that models will be generated from the factory NER instead of picking the trained NER. Running the assembled pipeline also assigns entity tags from the labels in a factory NER instead of the labels mentioned in the trained NER.
The `./model_store/assembled/config.cfg` looks something like this:

```
[paths]
train = null
dev = null
vectors = null
init_tok2vec = null

[system]
gpu_allocator = null
seed = 0

[nlp]
lang = "en"
pipeline = ["ner_1","ner_2"]
batch_size = 256
disabled = []
before_creation = null
after_creation = null
after_pipeline_creation = null
tokenizer = {"@tokenizers":"spacy.Tokenizer.v1"}

[components]

[components.ner_1]
factory = "ner"
moves = null
update_with_oracle_cut_size = 100

[components.ner_1.model]
@architectures = "spacy.TransitionBasedParser.v2"
state_type = "ner"
extra_state_tokens = false
hidden_width = 64
maxout_pieces = 2
use_upper = true
nO = null

[components.ner_1.model.tok2vec]
@architectures = "spacy.Tok2Vec.v1"

[components.ner_1.model.tok2vec.embed]
@architectures = "spacy.MultiHashEmbed.v1"
width = 96
attrs = ["NORM","PREFIX","SUFFIX","SHAPE"]
rows = [5000,2500,2500,2500]
include_static_vectors = true

[components.ner_1.model.tok2vec.encode]
@architectures = "spacy.MaxoutWindowEncoder.v1"
width = 96
depth = 4
window_size = 1
maxout_pieces = 3

[components.ner_2]
factory = "ner"
moves = null
update_with_oracle_cut_size = 100

[components.ner_2]
@architectures = "spacy.TransitionBasedParser.v2"
state_type = "ner"
extra_state_tokens = false
hidden_width = 64
maxout_pieces = 2
use_upper = true
nO = null

[components.ner_2.model.tok2vec]
@architectures = "spacy.HashEmbedCNN.v2"
pretrained_vectors = null
width = 96
depth = 4
embed_size = 2000
window_size = 1
maxout_pieces = 3
subword_features = true

[corpora]

[corpora.dev]
@readers = "spacy.Corpus.v1"
path = ${paths.dev}
gold_preproc = false
max_length = 0
limit = 0
augmenter = null

[corpora.train]
@readers = "spacy.Corpus.v1"
path = ${paths.train}
gold_preproc = false
max_length = 0
limit = 0
augmenter = null

[training]
seed = ${system.seed}
gpu_allocator = ${system.gpu_allocator}
dropout = 0.1
accumulate_gradient = 1
patience = 1600
max_epochs = 0
max_steps = 20000
eval_frequency = 200
frozen_components = []
dev_corpus = "corpora.dev"
train_corpus = "corpora.train"
before_to_disk = null

[training.batcher]
@batchers = "spacy.batch_by_words.v1"
discard_oversize = false
tolerance = 0.2
get_length = null

[training.batcher.size]
@schedules = "compounding.v1"
start = 100
stop = 1000
compound = 1.001
t = 0.0

[training.logger]
@loggers = "spacy.ConsoleLogger.v1"
progress_bar = false

[training.optimizer]
@optimizers = "Adam.v1"
beta1 = 0.9
beta2 = 0.999
L2_is_weight_decay = true
L2 = 0.01
grad_clip = 1.0
use_averages = false
eps = 0.00000001
learn_rate = 0.001

[training.score_weights]
ents_per_type = null
ents_f = 0.85
ents_p = 0.0
ents_r = 0.0
sents_f = 0.17
sents_p = 0.0
sents_r = 0.0

[pretraining]

[initialize]
vectors = null
init_tok2vec = null
vocab_data = null
lookups = null
before_init = null
after_init = null

[initialize.components]

[initialize.tokenizer]

[source]
ner_1 = "./model_store/ner_1/"
ner_2 = "./model_store/ner_2/"
```

## Your Environment
- **spaCy version:** 3.0.6
- **Platform:** macOS-10.16-x86_64-i386-64bit
- **Python version:** 3.8.8
- **Pipelines:** en_core_web_lg (3.0.0), en_core_web_md (3.0.0), en_core_web_sm (3.0.0), en_core_web_trf (3.0.0)

**Comment by adrianeboyd** [comment]
Hi, the assembled config looks correct: after you run `spacy assemble`, the `source = ` specification should be replaced with an exact copy of the config from the source pipeline and all the model weights/settings are also copied into the new pipeline. After this step, you don't need to have the local paths to the individual components available to load the pipeline -- all the data is copied directly into the new pipeline.

You should be able to see that the `[components.ner]` block looks the same in both `ner_#` and `assembled`. These aren't "factory" NER settings, it should be the exact settings from the sourced config. In the example above, you can see that the `ner_1` and `ner_2` configs in the assembled version have slightly different `tok2vec` settings (which is totally fine, of course!).

Can you double-check that the component annotations for `ner_1` in the original and combined pipelines are the same? For example, to check `ner_1`, compare:

```python
ner_1_nlp = spacy.load("path/ner_1")
assembled_nlp = spacy.load("path/assembled", disable=["ner_2"])
```

You can also just double-check the labels:

```python
ner_1_nlp = spacy.load("path/ner_1")
print(ner_1_nlp.get_pipe("ner_1").labels)
assembled_nlp = spacy.load("path/assembled", disable=["ner_2"])
print(assembed_nlp.get_pipe("ner_1").labels)
```

What you might be seeing is that NER components don't overwrite existing annotation, so if the predicted `ner_2` spans overlap with the predicted `ner_1` spans, there may not be as many `ner_2` entities as with `ner_2` by itself. The component that runs first has priority, so you might see different results with `ner_2, ner_1` vs. `ner_1, ner_2`.

**Reply by narayanacharya6** [reply]
Thanks for your response! The trained NER pipes do have the correct labels as I would expect. Also, I am aware of the span overlap constraint and use an ugly hack in-between the NER components from [here](https://github.com/explosion/spaCy/discussions/8223#discussioncomment-799096) to reset the `doc.ents`. 

But I believe you are right, the issue does not seem to be with the configs. While checking the labels in the `meta.json` file of the assembled as well as the individual trained models, I noticed that the vectors seem to be not as I would expect them(?).
```
"vectors":{
    "width":0,
    "vectors":0,
    "keys":0,
    "name":null
  }
```

One of the NERs in the assembled pipeline is simply a factory NER which has the below in it's `meta.json`:
```
"vectors":{
    "width":300,
    "vectors":20000,
    "keys":684830,
    "name":"en_vectors"
  }
```

I guess I need to investigate further in this direction.

**Reply by adrianeboyd** [reply]
Ah, the vectors are the issue here then, and it's something that's hard for `source` to detect automatically because there are so many ways `tok2vec` could include them.

I thought there should be a warning if the vectors don't match between the `source` and final pipeline, but it's possible something isn't working as intended, or you don't see the warning from `spacy assemble` for some reason. We can have another look at the warnings here.

To copy the vectors (which only seem to be used by `ner_1` from what I can tell?), you can specify `ner_1` as `[initialize.vectors]`, or you can use a callback to copy the whole vocab + tokenizer, if you have additional customizations beyond the plain `en`, in `[initialize.before_init]`:

https://spacy.io/api/top-level#copy_from_base_model

If you have vectors available, it might make sense to also use them for `ner_2`, too?

**Reply by narayanacharya6** [reply]
I did not see any warnings when I ran `assemble`. 

Also, looking at `spacy.cli.assemble.py` I see this:

https://github.com/explosion/spaCy/blob/master/spacy/cli/assemble.py#L48-L52

Will the initialize block be run for the sourced components?

**Reply by adrianeboyd** [reply]
The warning should come from this line, I think:

https://github.com/explosion/spaCy/blob/ff5cf3606cf354c7639ce40b81cacc93bfb03583/spacy/cli/assemble.py#L46

`spacy assemble` runs initialize for all components **except** the sourced components, since there may be rule-based components that load data in initialize. Sourced components should have already been initialized (and then trained, if relevant) in their original pipeline.

I'll have another look at how the warning shows up.

**Reply by adrianeboyd** [reply]
Ah, it is a bug. When it's loading the source models, it copies the vectors from that model into the vocab (this is a bug), so you don't see the warning because it looks okay temporarily, but then `ner_2` also copies its (0-length) vectors into the vocab, so the final model doesn't have them and you didn't see any warnings.

If you just source `ner_1`, it would currently work, but it's unintentional that the vectors are included just from `source`. The correct approach is to include them with `[initialize.vectors]`. If you do this, I think it should all work as intended.

**Reply by narayanacharya6** [reply]
So, I just added the `[initialize.vectors]` block to the configs of all the NERs that do not depend on vectors to ensure that the final assembled model does not get the 0-length vectors. I guess this helps side-step the issue for now. If you agree its a bug, should I raise an issue (or maybe you can raise one since you have more insight into the underlying problem :P )

If there were multiple NER models with different trained vectors, does that mean that the current implementation of `assemble` will allow only the last vocab's vectors in the final assembled model? (To be honest, I am slightly confused at the moment and not sure if the question even makes sense)

**Reply by adrianeboyd** [reply]
Hopefully #8388 will fix the vectors issue so that it shows the intended warnings and doesn't copy the vectors from the source models. `spacy assemble` should use the vectors from `[initialize.vectors]` instead of any vectors from the source models.

A spacy pipeline only supports one set of vectors, so it's not possible to source components from pipelines that use two different sets of vectors (easily*). You would need to retrain the your components so that they all use the same vectors.

(*It's not theoretically impossible to have each component have its own set of vectors with custom components, but the built-in components all use the vectors from `nlp.vocab.vectors` since the vector data is typically large and it's better to have one central source of vectors than to duplicate them in each component's saved model.)
